src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateActions = [];
+  const stateCopy = {
+    ...state,
+  };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clearProperties(stateCopy);
+        break;
+
+      default:
+        return `Error, unknown action type: ${action.type}`;
+    }
+    stateActions.push({ ...stateCopy });
+  }
+
+  return stateActions;
+}
+
+function addProperties(stateCopy, extraData) {
+  Object.assign(stateCopy, extraData);
+}
+
+function removeProperties(stateCopy, keysToRemove) {
+  for (const key of keysToRemove) {
+    delete stateCopy[key];
+  }
+}
```

Comment:
[NAMING] - Similar to the `addProperties` function, `removeProperties` also mutates the `stateCopy`. It would be more clear if the parameter was named to reflect that it's the current state, such as `currentState`.